### PR TITLE
Refactor Classes to Structs

### PR DESCRIPTION
**Pull request by teinarss**
image

### CODE REVIEW

**Comment by pchote** [comment]
Won't this trigger the usual set of mutable struct problems?

**Comment by teinarss** [comment]
I don't think so, took the idea from
https://github.com/dotnet/runtime/blob/8593a477fb8f41029f9d7963658922b8d504c76e/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs

**Comment by teinarss** [comment]
https://github.com/dotnet/runtime/blob/8593a477fb8f41029f9d7963658922b8d504c76e/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L1077

**Review by pchote — APPROVED** [review]

File: OpenRA.Game/Map/ProjectedCellRegion.cs [hunk 1]
```diff
@@ -86,9 +86,11 @@ public sealed class ProjectedCellRegionEnumerator : IEnumerator<PPos>
 			PPos current;
 
 			public ProjectedCellRegionEnumerator(ProjectedCellRegion region)
+				: this()
 			{
 				r = region;
 				Reset();
+				current = new PPos(u, v);
```
Comment (abcdefg30) [inline OpenRA.Game/Map/ProjectedCellRegion.cs:93]: `PPos.Zero`?

File: OpenRA.Game/Map/CellRegion.cs [hunk 2]
```diff
@@ -128,9 +128,11 @@ public sealed class CellRegionEnumerator : IEnumerator<CPos>
 			CPos current;
 
 			public CellRegionEnumerator(CellRegion region)
+				: this()
 			{
 				r = region;
 				Reset();
+				current = new MPos(u, v).ToCPos(r.gridType);
```
Comment (abcdefg30) [inline OpenRA.Game/Map/CellRegion.cs:135]: `CPos.Zero`? (`MPos.Zero.ToCPos` should be that.)

**Comment by abcdefg30** [comment]
Actually nvm, `Reset` changes `u` and `v`.

**Review by abcdefg30 — APPROVED** [review]
Ok.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9c44897856131ff145299a52f2accd1df67d9724...4803f69eba390d99e6a554f89a659cf866dcadc9)

**Pull request merged**